internal/core/processor/universal/updater/updater.go

```diff
@@ -0,0 +1,248 @@
+	ownershipContractEvoEventIndex, err := tx.GetCurrentEvoEventsIndexForOwnershipContract(contract)
+	if err != nil {
+		return 0, err
+	}
+	var transferIndex int
+	for {
+		switch {
+		// all events have been processed => return
+		case ownershipContractEvoEventIndex >= uint64(len(mintedEvents)) && transferIndex >= len(modelTransferEvents):
+			return ownershipContractEvoEventIndex, nil
+		// all minted events have been processed => process remaining transfer events
+		case ownershipContractEvoEventIndex >= uint64(len(mintedEvents)):
+			if err := updateStateWithTransfer(contract, tx, &modelTransferEvents[transferIndex]); err != nil {
+				return 0, err
+			}
+			transferIndex++
+		// all transfer events have been processed => process remaining minted events
+		case transferIndex >= len(modelTransferEvents):
+			if mintedEvents[ownershipContractEvoEventIndex].Timestamp < lastBlockTimestamp {
+				err := updateStateWithMint(ctx, client, contract, tx, &mintedEvents[ownershipContractEvoEventIndex])
+				if err != nil {
+					return 0, err
+				}
+				ownershipContractEvoEventIndex++
+			} else {
+				return ownershipContractEvoEventIndex, nil
+			}
+
+		default:
+			// if minted event's timestamp is behind transfer event's timestamp => process minted event
+			if mintedEvents[ownershipContractEvoEventIndex].Timestamp < modelTransferEvents[transferIndex].Timestamp {
+				err := updateStateWithMint(ctx, client, contract, tx, &mintedEvents[ownershipContractEvoEventIndex])
+				if err != nil {
+					return 0, err
+				}
+				ownershipContractEvoEventIndex++
+			} else {
+				if err := updateStateWithTransfer(contract, tx, &modelTransferEvents[transferIndex]); err != nil {
+					return 0, err
+				}
+				transferIndex++
+			}
+		}
+	}
+}
+
+func updateStateWithTransfer(contract string, tx state.Tx, transferEvent *model.ERC721Transfer) error {
+	slog.Debug("updating state with transfer event", "modelTransferEvent", transferEvent, "contract", contract)
+	err := tagRootsUntilBlock(tx, contract, transferEvent.BlockNumber)
+	if err != nil {
+		return err
+	}
+
+	return tx.Transfer(common.HexToAddress(contract), transferEvent)
+}
+
+func updateStateWithMint(ctx context.Context, client scan.EthClient, contract string, tx state.Tx, mintedEvent *model.MintedWithExternalURI) error {
+	slog.Debug("updating state with mint event", "modelTransferEvent", mintedEvent, "contract", contract)
+	block, err := getFirstOwnershipBlockAfterMintEvent(ctx, client, contract, tx, mintedEvent)
+	if err != nil {
+		return err
+	}
+
+	err = tagRootsUntilBlock(tx, contract, block)
+	if err != nil {
+		return err
+	}
+
+	return tx.Mint(common.HexToAddress(contract), mintedEvent)
+}
+
+func getFirstOwnershipBlockAfterMintEvent(ctx context.Context,
+	client scan.EthClient,
+	contract string,
+	tx state.Tx,
+	mintedEvent *model.MintedWithExternalURI,
+) (uint64, error) {
+	// TODO at this point this function uses GetLastTaggedBlock(). I think the first ownership block after mint event
```

Comment:
why do we keep this TODO?

Reply:
I will remove TODO part. I think this will going to be changed. probably when you do reorg handling.

for sure there we will have to have mapping evoBlock-ownershipBlock. Then this function will be just a call to that mapping.